### PR TITLE
Themes Selectors: Use theme ID instead of object as arg, take two

### PR DESCRIPTION
**Pull request by ockham**
Includes everything from #10434 and #11843 (which were reverted by #11846 and #11851, respectively).

For more information and testing instructions, see #10434 (and #11843).

Additionally, this PR fixes
* the signup issue which had us revert #10434 -- be sure to cover that when testing!
* the 'Customize' option for the currently active theme (test both a WP.com site, and an active theme in both the upper and lower list on a JP site).

~TODO~ Done:
* [x] Fix signup! Probably just a `getScreenshotOption` somewhere
  * [This function](https://github.com/Automattic/wp-calypso/blob/0b3fcce343d371d8f9ede83cbcd962bf6f56d234/client/signup/steps/theme-selection/index.jsx#L41) is [passed](https://github.com/Automattic/wp-calypso/blob/0b3fcce343d371d8f9ede83cbcd962bf6f56d234/client/signup/steps/theme-selection/index.jsx#L92) [as](https://github.com/Automattic/wp-calypso/blob/0b3fcce343d371d8f9ede83cbcd962bf6f56d234/client/signup/steps/theme-selection/signup-themes-list.jsx#L63) `getScreenshotOption`. With this PR, it will get a `themeId` instead of a `theme` obj. (FYI @kwight 😉 ) It will then need to look up the corresponding theme object from [that awful custom data structure](https://github.com/Automattic/wp-calypso/blob/0b3fcce343d371d8f9ede83cbcd962bf6f56d234/client/lib/signup/themes-data.js) (which should die in a 🔥  BTW).
* [x] Check if active theme's 'Customize' option works, fix otherwise

### CODE REVIEW

**Comment by kwight** [comment]
> It will then need to look up the corresponding theme object from that awful custom data structure

Yes, I imagined the fix would involve getting rid of that `themes-data.js` completely, and replacing [this import](https://github.com/Automattic/wp-calypso/blob/master/client/lib/signup/themes.js#L10) with selectors that fetch the needed themes from the store (all that `getThemes()` logic is still needed). The second part would be adding the data that's in `themes-data.js` but missing from a store theme: `fallback` and `design` (looks like `verticals` isn't used any more, but with marketing starting to develop signup, I'm sure it'll come up again).

**Comment by kwight** [comment]
Oh, and let me know if I can help at all :)

File: client/state/themes/selectors.js [hunk 1]
```diff
@@ -411,40 +411,44 @@ export function getThemeHelpUrl( state, theme, siteId ) {
...
+ * @param  {String}   themeId Theme ID
+ * @param  {?Number}  siteId  Site ID to open the customizer for
+ * @return {?String}          Customizer URL
  */
-export function getThemeCustomizeUrl( state, theme, siteId ) {
+export function getThemeCustomizeUrl( state, themeId, siteId ) {
 	const customizerUrl = getCustomizerUrl( state, siteId );
 
-	if ( ! ( siteId && theme ) ) {
+	if ( ! ( siteId && themeId ) || isThemeActive( state, themeId, siteId ) ) {
```
Comment (seear) [inline client/state/themes/selectors.js]: 👌

**Comment by folletto** [comment]
Checked on Chrome:

1. All Sites → List / Scroll — ✅ 
2. All Sites → Search string — ✅ 
3. All Sites → Free / Premium toggle — ✅ 
4. All Sites → ••• → Live Demo — ✅ 
5. All Sites → ••• → Activate — ✅ 
6. All Sites → ••• → Try & Customize — ✅ 
7. All Sites → ••• → Info / Support — ✅ 
8. Single Site → List / Scroll — ✅ 
9. Single Site → Search string — ✅ 
10. Single Site → Free / Premium toggle — ✅ 
11. Single Site → ••• → Live Demo — ❌ 
12. Single Site → ••• → Activate — ✅ 
13. Single Site → ••• → Try & Customize — ✅ 
14. Single Site → ••• → Info / Support — ✅ 
15. Direct → Sheet → Pick (Free) — ✅ 
16. Direct → Sheet → Pick (Premium) — ✅ 
17. Single Site → Sheet → Pick (Free) — ✅ 
18. Single Site → Sheet → Pick (Premium) — ❌ 
19. Registration Funnel → Theme showcase → theme list — ✅ 
20. Registration Funnel → Homepage → theme list — ✅ 
21. Registration Funnel → Full funnel + check frontend — ✅ 

#### (11) Single Site → ••• → Live Demo

The button doesn't work on theme "Perennial". Seems failing on Premium, but working on Free ones.

`TypeError: Cannot read property `getUrl` of undefined @ theme-preview.jsx:93` 

#### (18) Single Site → Sheet → Pick (Premium)

This happened:

screen shot 2017-03-10 at 13 20 30

Clicking again then... worked. Super odd.

**Comment by ockham** [comment]
@folletto Thanks for testing!

#### (11) Single Site → ••• → Live Demo

Should be fixed now.

#### (18) Single Site → Sheet → Pick (Premium)

Aw, that looks hard to repro (I couldn't for now). The error in the screenshot clearly looks like an API request that's gone wrong (`https://public-api.wordpress.com/rest/v1undefined?http_envelope=1`), but it seems to come from jQuery, and we don't use that, so it might be unrelated. (Unless the actual error is further up and not visible in the screenshot.)

Anyway, I seem to recall there were issues with purchase in `master`, too (but fortunately not in production) -- can you check that?

**Comment by budzanowski** [comment]
This is fantastic piece of work! I was also not able to reproduce the api error -  You are right looks very unrelated.

**Comment by folletto** [comment]
> `https://public-api.wordpress.com/rest/v1undefined?http_envelope=1`

This is unrelated. I get it all the times everywhere.

There were no errors in the console related to that.

**Pull request merged**